astropy/modeling/core.py

```diff
@@ -1224,8 +1392,114 @@ def _param_sets(self, raw=False):
             psets[:] = values
             return psets
 
+        # TODO: Returning an array from this method may be entirely pointless
+        # for internal use--perhaps only the external param_sets method should
+        # return an array (and just for backwards compat--I would prefer to
+        # maybe deprecate that method)
+
         return np.array(values)
 
+    def _prepare_input_units(self, inputs):
+        using_quantities = self._using_quantities
+        using_quantities |= any(isinstance(input_, Quantity)
+                                for input_ in inputs)
+
+        if using_quantities:
+            inputs = self._prepare_variable_units(inputs)
+
+        return inputs, using_quantities
+
+    def _prepare_output_units(self, inputs, outputs):
+        if not self._using_quantities:
+            return outputs
+        else:
+            return self._prepare_variable_units(inputs, outputs)
+
+    def _prepare_variable_units(self, inputs, outputs=None):
+        if outputs is None:
+            unit_specs = self.input_units
+            var_names = self.inputs
+            n_vars = self.n_inputs
+            vars_ = inputs
+        else:
+            unit_specs = self.output_units
+            var_names = self.outputs
+            n_vars = self.n_outputs
+            vars_ = outputs
+
+        if unit_specs is None:
```

Comment:
Tiny, but this could be
```
if not isinstance(unit_specs, tuple):
```
or, if speed is important
```
if unit_specs is None or not isinstance(unit_specs, tuple):
```